webdriver/appserver.js

```diff
@@ -0,0 +1,98 @@
+/**
+ * Copyright 2019 The WPT Dashboard Project. All rights reserved.
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ */
+
+const { spawn } = require('child_process');
+const { debug } = require('debug');
+const process = require('process');
+
+const log = debug('wpt.fyi');
+
+const ready = Symbol('ready');
+
+class DevAppserver {
+  /**
+   * @typedef {object} DevAppserverConfig
+   * @property {string} project
+   * @property {number} port
+   * @property {number} gcdPort
+   */
+  /**
+   * @param {DevAppserverConfig} config
+   */
+  constructor(config) {
+    this.config = Object.freeze(
+      Object.assign({
+        project: 'test-app',
+        port: 9090,
+        gcdPort: 9091,
+      }, config)
+    );
+    this.process = startDevAppserver(this.config);
+    this[ready] = this._awaitReady(this.process);
+
+    /**
+     * @type {URL} url The URL of the dev_appserver frontend
+     */
+    this.url = `http://localhost:${this.config.port}`;
+  }
+
+  /**
+   * @type {Promise} ready
+   */
+  get ready() {
+    return this[ready];
+  }
+
+  _awaitReady(process) {
+    return new Promise(resolve => {
+      const logDevAppserver = debug('wpt.fyi:appserver');
+      process.stderr.on('data', buffer => {
+        logDevAppserver(buffer.toString());
+      });
+
+      // TODO(Hexcles): do we still need to check if the server is up?
```

Comment:
Reminder here in case you forget

Reply:
Thanks. I'll leave this TODO as is for now. The server starts almost instantaneously.